### PR TITLE
[filebeat][gcs] - Added support for new features and removed partial save mechanism

### LINKED ISSUE #35847: Improvements to the gcs input
1) Add support for bucket prefix similar to awss3 input.
2) Add support for expand event from field feature similar to awss3
3) Add support for date filtering.
4) Remove partial save feature because of concurrency issues.

### PR DESCRIPTION
**Pull request by ShourieG**
## Type of change
- Enhancement
- Docs

## Proposed commit message
Added support for the following features as part of a planned enhancement update: 
1)  File selectors support similar to aws-s3 input, but only supports regex matching atm.
2) A date filter for blobs based on unix epoch (seconds) format.
3) Expand event list from field similar to aws-s3 input.
4) Removed partial save mechanism for now due to concurrency issues.

These additions help make the input more robust for further use. 
## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

- Closes #35847

### CODE REVIEW

File: x-pack/filebeat/input/gcs/job.go
Comment (bhapas) [inline x-pack/filebeat/input/gcs/job.go:160]: Nice to know why [UseNumber](https://pkg.go.dev/encoding/json#Decoder.UseNumber) here?

File: x-pack/filebeat/input/gcs/job.go
Comment (bhapas) [inline x-pack/filebeat/input/gcs/job.go:196]: A more detailed error message may be.. This is particularly when `publishing` , so something which tells that?

File: x-pack/filebeat/input/gcs/job.go
Comment (bhapas) [inline x-pack/filebeat/input/gcs/job.go:202]: Same error message comment here

File: x-pack/filebeat/input/gcs/job.go
Comment (bhapas) [inline x-pack/filebeat/input/gcs/job.go:213]: Formatted error suggesting an `Unmarshal` error

File: x-pack/filebeat/input/gcs/job.go
Comment (bhapas) [inline x-pack/filebeat/input/gcs/job.go:226]: Formatted error suggesting `Token` error

File: x-pack/filebeat/input/gcs/job.go
Comment (bhapas) [inline x-pack/filebeat/input/gcs/job.go:243]: Formatted error suggesting an `Marshal` error

File: x-pack/filebeat/input/gcs/job.go
Comment (bhapas) [inline x-pack/filebeat/input/gcs/job.go:251]: Same comment for error message regarding `publish`

File: x-pack/filebeat/input/gcs/job.go
Comment (bhapas) [inline x-pack/filebeat/input/gcs/job.go:257]: Same comment for error message regarding `publish`

File: x-pack/filebeat/input/gcs/scheduler.go
Comment (bhapas) [inline x-pack/filebeat/input/gcs/scheduler.go:251]: Is it possible to configure a `nil` in `file_selectors regex`

File: x-pack/filebeat/docs/inputs/input-gcs.asciidoc
Comment (ShourieG) [reply x-pack/filebeat/docs/inputs/input-gcs.asciidoc:82]: Before creating an issue here, we need to do some research on our end to figure out the bottlenecks, which atm are proving to be quite tricky.

Comment (bhapas) [reply x-pack/filebeat/docs/inputs/input-gcs.asciidoc:82]: Yeah. This can be more of a tracking issue which can have followup technical issues

File: x-pack/filebeat/input/gcs/input_test.go
Comment (ShourieG) [reply x-pack/filebeat/input/gcs/input_test.go:482]: We cannot have empty buckets or file selectors as they would result in panics as these attributes are expected to have values if they are defined, however we can have an empty regex value. (non-nil empty string). Adding a test for that.

File: x-pack/filebeat/input/gcs/scheduler.go
Comment (ShourieG) [reply x-pack/filebeat/input/gcs/scheduler.go:251]: not possible to configure nil, but we can have an empty string. This is just a safety net check.

Comment (bhapas) [reply x-pack/filebeat/input/gcs/scheduler.go:251]: But this line gives a wrong impression that if you [some how] manage to get a `nil` in regex then all file patterns are allowed. Doesn't it?

And what happens when regex is `empty string` , just to understand.

Comment (ShourieG) [reply x-pack/filebeat/input/gcs/scheduler.go:251]: if regex is an empty string, it will behave like a no-op.

Comment (bhapas) [reply x-pack/filebeat/input/gcs/scheduler.go:251]: Great.. Then , we can safely remove the `nil check`, can't we?

Comment (ShourieG) [reply x-pack/filebeat/input/gcs/scheduler.go:251]: yea

Comment (bhapas) [reply x-pack/filebeat/input/gcs/scheduler.go:251]: This is more on a `readability` point of view for me as a developer.. Just not to get mislead looking at a `== nil` always matches the file selector. :-)

Comment (ShourieG) [reply x-pack/filebeat/input/gcs/scheduler.go]: removed the check

File: x-pack/filebeat/docs/inputs/input-gcs.asciidoc
Comment (ShourieG) [reply x-pack/filebeat/docs/inputs/input-gcs.asciidoc:82]: Issue created - https://github.com/elastic/beats/issues/36717

File: x-pack/filebeat/input/gcs/config.go
Comment (ShourieG) [reply x-pack/filebeat/input/gcs/config.go:58]: issue created - https://github.com/elastic/beats/issues/36718

**Comment by ShourieG** [comment]
@bhapas could you approve if all else looks ok. Addressed all the comments, and have set the pr to auto-merge as I will be going out now.

**Comment by bhapas** [comment]
> @bhapas is it ok for an approval now ?

@ShourieG LGTM. But wanted a second review from an experienced dev as this is a big change

**Comment by ShourieG** [comment]
@bhapas I also agree, but I'm worried it might miss FF. Can we keep it approved, while I look for a senior dev to review this. It similar to the azure blob pr which was reviewed by Dan and Andrew.

**Comment by ShourieG** [comment]
@bhapas I also agree, but I'm worried it might miss FF. Can we keep it approved, while I look for a senior dev to review this. It similar to the azure blob pr which was reviewed by Dan and Andrew.

**Comment by ShourieG** [comment]
@andrewkroh, could you give a quick review for this PR, trying to meet FF for tomorrow.

File: x-pack/filebeat/docs/inputs/input-gcs.asciidoc
Comment (ShourieG) [reply x-pack/filebeat/docs/inputs/input-gcs.asciidoc:279]: This decision was made in line with the implementation in aws-s3 input. This enables us to expand the fileselector implementation in future and have combinations of regex's and expand event from list ops and custom decoders etc.

File: x-pack/filebeat/docs/inputs/input-gcs.asciidoc
Comment (ShourieG) [reply x-pack/filebeat/docs/inputs/input-gcs.asciidoc]: removed the note.

**Comment by ShourieG** [comment]
@efd6, Have resolved the PR suggestions.

File: x-pack/filebeat/input/gcs/scheduler.go
Comment (efd6) [inline x-pack/filebeat/input/gcs/scheduler.go:146]: Why is this done?

File: x-pack/filebeat/input/gcs/job.go
Comment (efd6) [reply x-pack/filebeat/input/gcs/job.go:161]: I'm not convinced a comment is needed since this is an idiom.

File: x-pack/filebeat/input/gcs/job.go
Comment (efd6) [inline x-pack/filebeat/input/gcs/job.go:201]: This is the only thing that makes me concerned here. When dealing with locks, "should" is a word that makes me feel uncomfortable. Words I like are "will" and "must", but only when 100% certain.

**Review by efd6 — APPROVED** [review]
LGTM

Comments are informational only; no action is required.

File: x-pack/filebeat/input/gcs/job.go
Comment (ShourieG) [reply x-pack/filebeat/input/gcs/job.go:201]: Yes actually this should have been worded better. There's no chance of any concurrency issues happening here since we don't update any shared resources.

File: x-pack/filebeat/input/gcs/scheduler.go
Comment (ShourieG) [reply x-pack/filebeat/input/gcs/scheduler.go:146]: Just so that unnecessary calls are not made to the method for the regex match if file selectors are not defined.

Comment (efd6) [reply x-pack/filebeat/input/gcs/scheduler.go:146]: `len` is not really a call.

**Pull request merged**
